Add isInputPending guide (#4220)
GoogleChrome/web.dev: src/site/content/en/blog/isinputpending/index.md

---

> title: Better JS scheduling with isInputPending()

Author: MDN's convention is to always end function/method names with parentheses

subhead: >

>   A new JavaScript API that may help you avoid the tradeoff between
>   load performance and input responsiveness.
> description: >
>   A new JavaScript API that may help you avoid the tradeoff between

Author: ```suggestion
  A new JavaScript API that may help you avoid the trade-off between
  load performance and input responsiveness.
description: >
  A new JavaScript API that may help you avoid the trade-off between
```

  load performance and input responsiveness.
date: 2020-11-13
hero: hero.jpg
alt: A traffic sign that says 'GIVE WAY'.
tags:
  - blog
  - javascript
  - performance
  - rendering
authors:
  - nateschloss
  - andrewcomminos
---

Loading fast is hard. Sites that leverage JS to render their content currently

> have to make a trade-off- either perform all the work needed for display all at
> once, or chunk the work into smaller tasks in order to remain responsive to
> input and paint -- occasionally, at the expense of performance.

Author: ```suggestion
have to make a trade-off between load performance and input 
responsiveness: either perform all the work needed for display
all at once (better load performance, worse input responsiveness), or
chunk the work into smaller tasks in order to remain responsive to 
input and paint (worse load performance, better input 
responsiveness).
```

Rephrasing to make the trade-off clearer

To eliminate the need to make this trade-off, Facebook proposed and implemented

> the `isInputPending()` API in Chrome in order to improve responsiveness without

Author: ```suggestion
the `isInputPending()` API in Chromium in order to improve responsiveness without
```

Unless this is actually only for Chrome and not for other Chromium-based browsers, we should use Chromium when referring to the actual browser (we can use Chrome to refer to the team that you collaborated with)

yielding. Based on origin trial feedback, we've made a number of updates to the

> API, and are happy to announce that the API is now shipping by default in Chrome

Author: ```suggestion
API, and are happy to announce that the API is now shipping by default in Chromium
```

> 87!
> 
> ## Background

Author: ```suggestion
87!

## Browser compatibility

`isInputPending()` is shipping in Chromium-based browsers starting in version 87.

## Background
```

We know that browser compatibility is a big concern for developers (https://web.dev/handbook/content-checklist/#compatibility). We should add a browser compat section to the start of the page as a courtesy. What's the status in other browsers? We usually either say "no signal" or "browser X is interested in implementing" or "browser X is implementing" or "browser X has shipped"

{% Aside %}
  For the full background, check out our Facebook Engineering blog post,
  [Faster input events with Facebook's first browser API 
  contribution](https://engineering.fb.com/developer-tools/isinputpending-api/).
{% endAside %}

Most work in today's JS ecosystem gets done on a single thread: the main thread.

> This provides a robust execution model to developers, but the user experience
> can suffer drastically if script executes for a long time, responsiveness
> in particular. If the page is doing a lot of work while an input event is fired,

Author: ```suggestion
This provides a robust execution model to developers, but the user experience
(responsiveness in particular) can suffer drastically if script executes for a long 
time. If the page is doing a lot of work while an input event is fired,
```
Rephrasing an awkward sentence

the browser to check its input event queue, process events, and pick up the next
JavaScript block. While the browser responds to events quicker, the overall
loading time of the page gets slowed down. And if we yield too often, the page
loads too slowly. If we yield less often, it takes longer for the browser to
respond to user events, and people get frustrated. Not fun.

![A diagram showing that when you run long JS tasks, the browser has less time
to dispatch events.](diagram.png)

At Facebook, we wanted to see what things would look like if we came up with a
new approach for loading that would eliminate this frustrating trade-off. We
reached out to our friends at Chrome about this, and came up with the proposal
for `isInputPending()`. The `isInputPending()` API is the first to use the concept of
interrupts for user inputs on the web, and allows for JavaScript to be
able to check for input without yielding to the browser.

![A diagram showing that isInputPending() allows your JS to check if there's
pending user input, without completely yielding execution back to the
browser.](legend.png)

Since there was interest in the API, we partnered with our colleagues at Chrome

> to implement and ship the feature in Chrome. With help from the Chrome

Author: ```suggestion
to implement and ship the feature in Chromium. With help from the Chrome
```

process input events, such as to perform rendering and execute other scripts on
the page.

There exist cases where the browser isn't able to properly attribute pending
input events. In particular, setting complex clips and masks for cross-origin
iframes may report false negatives (i.e. `isInputPending()` may unexpectedly return
false when targeting these frames). Be sure that you're yielding often enough if
your site does require interactions with stylized subframes.

Be mindful of other pages that share an event loop, as well. On platforms such
as Chrome for Android, it's quite common for multiple origins to share an event
loop. `isInputPending()` will never return `true` if input is dispatched to a
cross-origin frame, and thus backgrounded pages may interfere with the
responsiveness of foreground pages. You may wish to reduce, postpone, or yield
more often when doing work in the background using the [Page Visibility
API](https://www.w3.org/TR/page-visibility-2/).

We encourage you to use `isInputPending()` with discretion. If there isn't
user-blocking work to be done, then be kind to others on the event loop by
yielding more frequently. [Long tasks can be

> harmful](/long-tasks-devtools/).
> 
> ## Conclusion

Author: ```suggestion
harmful](/long-tasks-devtools/).

## Feedback

* TODO

## Conclusion
```

Can developers contact you two on Twitter? Does the spec have a repo?

Reviewer: Yes to all of the above! Twitter links above, the repo can be found at https://github.com/WICG/is-input-pending.

---
title: Better JS scheduling with isInputPending()
subhead: >
  A new JavaScript API that may help you avoid the trade-off between
  load performance and input responsiveness.
description: >
  A new JavaScript API that may help you avoid the trade-off between
  load performance and input responsiveness.

> date: 2020-11-13

Author: ```suggestion
date: 2020-11-19
```

W3C Web Performance Working Group and implemented changes to the API.

## Example: a yieldier scheduler

Suppose that you've got a bunch of display-blocking work to do to load your
page, for example generating markup from components, factoring out primes, or
just drawing a cool loading spinner. Each one of these is broken into a discrete
work item. Using the scheduler pattern, let's sketch out how we might process
our work in a hypothetical `processWorkQueue()` function:

```js
const DEADLINE = performance.now() + QUANTUM;
while (workQueue.length > 0) {
  if (performance.now() >= DEADLINE) {
    // Yield the event loop if we're out of time.
    setTimeout(processWorkQueue);
    return;
  }
  let job = workQueue.shift();
  job.execute();
}
```

By invoking `processWorkQueue()` later in a new macrotask via `setTimeout()`, we
give the browser the ability to remain somewhat responsive to input (it can
run event handlers before work resumes) while still managing to run relatively
uninterrupted. Though, we might get descheduled for a long time by other work
that wants control of the event loop, or get up to an extra `QUANTUM` milliseconds
of event latency.

> 

Reviewer: ```suggestion

A good value for `QUANTUM` (under the [RAIL model](/rail/)) is <50ms, depending on the type of work being done. This value is primarily what dictates the tradeoff between throughput and latency.

```

---
title: Better JS scheduling with isInputPending()
subhead: >
  A new JavaScript API that may help you avoid the trade-off between
  load performance and input responsiveness.
description: >
  A new JavaScript API that may help you avoid the trade-off between
  load performance and input responsiveness.
date: 2020-11-19
hero: hero.jpg
alt: A traffic sign that says 'GIVE WAY'.
tags:
  - blog
  - javascript
  - performance
  - rendering
authors:
  - nateschloss

>   - andrewcomminos
> ---

Author: ```suggestion
  - andrewcomminos
feedback:
  - api
---
```
